extra/innochecksum.cc

```diff
@@ -1428,6 +1453,101 @@ rewrite_checksum(
 		&& !write_file(filename, fil_in, buf, flags, pos);
 }
 
+/** Read and validate page 0, then initialize tablespace flags
+and page size.
+@param	fil_in        File pointer
+@param	buf           Buffer to read page into
+@return whether the page was read successfully */
+static bool read_and_validate_page0(FILE *fil_in, byte *buf)
+{
+  /* Read the minimum page size first */
+  size_t initial_page_size= UNIV_ZIP_SIZE_MIN;
+  if (use_tablespace_flags)
+  {
+    if (!fil_space_t::is_valid_flags(tablespace_flags,
+                                     cur_space != 0))
+    {
+      fprintf(stderr, "Warning: Provided tablespace_flags "
+              "is not valid. So skipping --tablespace_flags option");
+      use_tablespace_flags= false;
+    }
+    else
+    {
+      init_page_size_from_flags(tablespace_flags);
+      initial_page_size= physical_page_size;
+    }
+  }
+
+  size_t bytes= fread(buf, 1, initial_page_size, fil_in);
+
+  if (bytes != initial_page_size)
+  {
+    fprintf(stderr, "Error: Was not able to read the "
+            "minimum page size of %zu bytes. Bytes read "
+            "was %zu\n", initial_page_size, bytes);
+    return false;
+  }
+
+  /* Read space_id and page offset */
+  cur_space= mach_read_from_4(buf + FIL_PAGE_SPACE_ID);
+  cur_page_num= mach_read_from_4(buf + FIL_PAGE_OFFSET);
+
+  /* Get tablespace flags from the FSP header */
+  uint32_t flags= mach_read_from_4(buf + FSP_HEADER_OFFSET +
+                                   FSP_SPACE_FLAGS);
+
+  if (use_tablespace_flags)
+  {
+    if (cur_page_num == 0 && flags != tablespace_flags)
+    {
+      if (is_log_enabled)
+        fprintf(log_file,
+                "Fail: Mismatch between provided tablespace flags "
+                "and file flags\n");
+      fprintf(stderr, "Error: Mismatch between provided tablespace "
+              "flags (0x%x) and file flags (0x%x)\n",
+              tablespace_flags, flags);
+      return false;
+    }
+  }
+  else
+  {
+    if (cur_page_num)
+    {
+      fprintf(stderr, "Error: First page of the tablespace file "
+              "should be 0, but encountered page number %d.\n"
+              "If you are checking multi file system "
+              "tablespace files, please specify the correct "
+              "tablespace flags using --tablespace-flags option.\n",
+              cur_page_num);
```

Comment:
`"%d"` is incorrect for `uint32_t`. Please use `"%" PRIu32` so that this will work on page numbers that are larger than `1<<31`. It would be good to cover this in a test case. (Create a fake page with a valid CRC-32C and test it with and without the option.)